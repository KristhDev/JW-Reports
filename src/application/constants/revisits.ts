export const revisitsMessages = {
    ABOUT_MIN_LENGTH: 'La información de la persona debe tener al menos 10 caracteres.',
    ABOUT_REQUIRED: 'La información de la persona es requerida.',
    ADDED_SUCCESS: 'Ha agregado la revisita correctamente.',
    ADDRESS_MIN_LENGTH: 'La dirección debe tener al menos 10 caracteres.',
    ADDRESS_REQUIRED: 'La dirección es requerida.',
    COMPLETED_SUCCESS: 'Ha marcado como completa su revisita correctamente.',
    DELETED_SUCCESS: 'Ha eliminado la revisita correctamente.',
    EXPORTED_SUCCESS: 'Ha exportado sus revisitas correctamente. El archivo se encuentra en la carpeta de descargas de su dispositivo.',
    NEXT_VISIT_REQUIRED: 'La fecha de la último visita no puede estar vacía.',
    PERSON_MIN_LENGTH: 'El nombre de la persona debe tener al menos 2 caracteres.',
    PERSON_NAME_REQUIRED: 'El nombre de la persona es requerido.',
    UNSELECTED_COMPLETE: 'No hay una revisita seleccionada para completar.',
    UNSELECTED_DELETE: 'No hay una revisita seleccionada para eliminar.',
    UNSELECTED_UPDATE: 'No hay una revisita seleccionada para actualizar.',
    UPDATED_SUCCESS: 'Ha actualizado la revisita correctamente.',
}