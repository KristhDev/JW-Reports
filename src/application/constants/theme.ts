import { ItemOption } from '../../infrasturcture/interfaces';

/**
 * The theme options to select
 */
export const THEME_OPTIONS: ItemOption[] = [
    { label: 'Modo predeterminado', value: 'default' },
    { label: 'Modo claro', value: 'light' },
    { label: 'Modo oscuro', value: 'dark' }
];