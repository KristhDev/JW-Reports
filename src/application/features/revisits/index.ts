export {
    default as revisitsReducer,
    addRevisit,
    addRevisits,
    clearRevisits,
    INIT_REVISIT,
    removeRevisit,
    removeRevisits,
    REVISITS_INITIAL_STATE,
    setHasMoreRevisits,
    setIsLastRevisitLoading,
    setIsRevisitDeleting,
    setIsRevisitsExporting,
    setIsRevisitLoading,
    setIsRevisitsLoading,
    setLastRevisit,
    setRefreshRevisits,
    setRevisitFilter,
    setRevisits,
    setRevisitsPagination,
    setRevisitsScreenHistory,
    setSelectedRevisit,
    updateRevisit
} from './slice';

export * from './types';