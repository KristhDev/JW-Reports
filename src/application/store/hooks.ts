import { useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();