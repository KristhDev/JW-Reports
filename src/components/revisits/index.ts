export { RevisitCard } from './RevisitCard';
export { RevisitForm } from './RevisitForm';