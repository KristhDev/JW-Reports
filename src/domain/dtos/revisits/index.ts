export * from './complete.dto';
export * from './create.dto';
export * from './update.dto';