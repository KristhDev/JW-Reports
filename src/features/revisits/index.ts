export {
    default as revisitsReducer,
    INIT_REVISIT,
    addRevisit,
    addRevisits,
    clearRevisits,
    removeRevisit,
    removeRevisits,
    setHasMoreRevisits,
    setIsRevisitDeleting,
    setIsRevisitLoading,
    setIsRevisitsLoading,
    setRefreshRevisits,
    setRevisitFilter,
    setRevisits,
    setRevisitsPagination,
    setRevisitsScreenHistory,
    setSelectedRevisit,
    updateRevisit
} from './slice';