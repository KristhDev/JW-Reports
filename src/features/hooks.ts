import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;