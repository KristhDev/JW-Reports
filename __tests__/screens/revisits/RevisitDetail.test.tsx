import React from 'react';
import { render, screen } from '@testing-library/react-native';
import dayjs from 'dayjs';

/* Screens */
import { RevisitDetail } from '../../../src/screens/revisits';

/* Hooks */
import { useRevisits, useStatus, useTheme } from '../../../src/hooks';

/* Theme */
import { darkColors } from '../../../src/theme';

/* Mocks */
import { selectedRevisitStateMock, setSelectedRevisitMock } from '../../mocks';

const revisitPhoto = 'https://img.freepik.com/free-vector/nature-scene-with-river-hills-forest-mountain-landscape-flat-cartoon-style-illustration_1150-37326.jpg';

/* Mock hooks */
jest.mock('../../../src/hooks/useRevisits.ts');
jest.mock('../../../src/hooks/useStatus.ts');
jest.mock('../../../src/hooks/useTheme.ts');

describe('Test in <RevisitDetail /> screen', () => {
    (useRevisits as jest.Mock).mockReturnValue({
        state: selectedRevisitStateMock,
        completeRevisit: jest.fn(),
        saveRevisit: jest.fn(),
        setSelectedRevisit: setSelectedRevisitMock,
    });

    (useStatus as jest.Mock).mockReturnValue({
        setErrorForm: jest.fn(),
    });

    (useTheme as jest.Mock).mockReturnValue({
        state: { colors: darkColors }
    });

    beforeEach(() => {
        render(<RevisitDetail />);
        jest.clearAllMocks();
    });

    it('should to match snapshot', () => {
        expect(screen.toJSON()).toMatchSnapshot();
    });

    it('should render revisit date', () => {
        const selectedRevisit = selectedRevisitStateMock.selectedRevisit;

        /* Get elements with data of revisit */
        const title = screen.getByTestId('title-text');
        const aboutSection = screen.getByTestId('revisit-detail-about-section');
        const addressSection = screen.getByTestId('revisit-detail-address-section');
        const createdDate = screen.getByTestId('revisit-detail-created-date');

        /* Check if title exists and contain respective values */
        expect(title).toBeTruthy();
        expect(title.props.children).toBe(selectedRevisit.personName.toUpperCase());

        /* Check if about section exists and contain respective value */
        expect(aboutSection).toBeTruthy();
        expect(aboutSection.props.children[0].props.children.join('')).toBe(`Información de ${ selectedRevisit.personName }:`);
        expect(aboutSection.props.children[1].props.children).toBe(selectedRevisit.about);

        /* Check if address section exists and contain respective value */
        expect(addressSection).toBeTruthy();
        expect(addressSection.props.children[0].props.children).toBe('Dirección:');
        expect(addressSection.props.children[1].props.children).toBe(selectedRevisit.address);

        /* Check if date exists and contain respective value */
        expect(createdDate).toBeTruthy();
        expect(createdDate.props.children).toBe(dayjs(selectedRevisit.createdAt).format('DD/MM/YYYY'));
    });

    it('should render next visit when selectedRevisit.done is false', () => {
        const nextVisit = dayjs(selectedRevisitStateMock.selectedRevisit.nextVisit);

        /* Get next visit text */
        const nextVisitText = screen.getByTestId('revisit-detail-next-visit');

        /* Check if next visit text exists and contain respective value */
        expect(nextVisitText).toBeTruthy();
        expect(nextVisitText.props.children).toBe(` ${ nextVisit.format('DD') } de ${ nextVisit.format('MMMM') } del ${ nextVisit.format('YYYY') }`);
    });

    it('should render revisit again section when selectedRevisit.done is true', () => {

        /* Mock data of useRevisits */
        (useRevisits as jest.Mock).mockReturnValue({
            state: {
                ...selectedRevisitStateMock,
                selectedRevisit: {
                    ...selectedRevisitStateMock.selectedRevisit,
                    done: true
                }
            },
            completeRevisit: jest.fn(),
            saveRevisit: jest.fn(),
            setSelectedRevisit: setSelectedRevisitMock,
        });

        render(<RevisitDetail />);

        /* Get again section and check if exists */
        const revisitAgainSection = screen.getByTestId('revisit-detail-revisit-again-section');
        expect(revisitAgainSection).toBeTruthy();
    });

    it('should call image if exisit in revisit', () => {

        /* Mock data of useRevisits */
        (useRevisits as jest.Mock).mockReturnValue({
            state: {
                ...selectedRevisitStateMock,
                selectedRevisit: {
                    ...selectedRevisitStateMock.selectedRevisit,
                    photo: revisitPhoto
                }
            },
            completeRevisit: jest.fn(),
            saveRevisit: jest.fn(),
            setSelectedRevisit: setSelectedRevisitMock,
        });

        render(<RevisitDetail />);

        const selectedRevisit = {
            ...selectedRevisitStateMock.selectedRevisit,
            photo: revisitPhoto
        }

        /* Get image and text image */
        const photoImage = screen.getByTestId('revisit-detail-photo-image');
        const photoText = screen.getByTestId('revisit-detail-photo-text');

        /* Check if image photo and image text exists and contain respective value */
        expect(photoImage).toBeTruthy();
        expect(photoImage.props.source.uri).toBe(revisitPhoto);
        expect(photoText).toBeTruthy();
        expect(photoText.props.children.join('')).toBe(`La foto es para ayudarte a recordar el lugar de residencia de ${ selectedRevisit.personName }`);
    });

    it('should not call setSelectedRevisit when index of navigation is different of 0', () => {

        /* Check if setSelectedRevisit isnt called */
        expect(setSelectedRevisitMock).not.toHaveBeenCalled();
    });
});