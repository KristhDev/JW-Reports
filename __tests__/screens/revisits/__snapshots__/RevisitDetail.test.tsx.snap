// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <RevistDetail /> screen should to match snapshot 1`] = `
[
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "center",
        "flexGrow": 1,
        "padding": 24,
        "paddingBottom": 100,
      }
    }
    overScrollMode="never"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "alignItems": "flex-start",
              "marginBottom": 40,
              "width": "100%",
            },
            {
              "paddingBottom": 0,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#FFFFFF",
                "fontSize": 38,
                "fontWeight": "bold",
              },
              {
                "fontSize": 24,
              },
            ]
          }
          testID="title-text"
        >
          DARIANA HAND
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 32,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontSize": 19,
                "fontWeight": "bold",
                "marginBottom": 0,
              }
            }
          >
            Próxima visita:
          </Text>
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontSize": 19,
              }
            }
            testID="revisit-detail-next-visit"
          >
             09 de December del 2021
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#A0B9E2",
                "fontSize": 19,
              }
            }
          >
            ¿Ya la visitaste?
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginBottom": 32,
            "width": "100%",
          }
        }
        testID="revisit-detail-about-section"
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 16,
            }
          }
        >
          Información de 
          Dariana Hand
          :
        </Text>
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
            }
          }
        >
          Ratione quae debitis pariatur tempora magnam est voluptatibus. Eveniet ipsum omnis maxime et soluta iusto. Autem aut ducimus dolore et recusandae et inventore. Enim autem in hic optio.
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 32,
            "width": "100%",
          }
        }
        testID="revisit-detail-address-section"
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 16,
            }
          }
        >
          Dirección:
        </Text>
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
            }
          }
        >
          Voluptatem ratione velit officia qui non ab aut atque. Voluptate possimus dolor ex commodi aut numquam voluptatem sit. Molestiae ut tempora unde sapiente. Soluta culpa repellat et voluptatibus odit eveniet dolore harum.
        </Text>
      </View>
      <Text
        style={
          {
            "bottom": 24,
            "color": "#B4B4B4",
            "fontSize": 16,
            "position": "absolute",
            "right": 24,
          }
        }
        testID="revisit-detail-created-date"
      >
        09/12/2021
      </Text>
    </View>
  </RCTScrollView>,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  />,
]
`;
